echo-notifications/src/main/groovy/com/netflix/spinnaker/echo/slack/SlackNotificationService.groovy

```diff
@@ -58,12 +82,95 @@ class SlackNotificationService implements NotificationService {
       if (sendCompactMessages) {
         response = slack.sendCompactMessage(token, new CompactSlackMessage(text), address, true)
       } else {
-        response = slack.sendMessage(token, new SlackAttachment("Spinnaker Notification", text), address, true)
+        response = slack.sendMessage(token,
+          new SlackAttachment("Spinnaker Notification", text, notification.interactiveActions),
+          address, true)
       }
       log.trace("Received response from Slack: {} {} for message '{}'. {}",
         response?.status, response?.reason, text, response?.body)
     }
 
     new EchoResponse.Void()
   }
+
+  private Map parseSlackPayload(Map content) {
+    if (!content.payload) {
+      throw new IllegalArgumentException("Missing payload field in Slack callback request.")
+    }
+
+    Map payload = objectMapper.readValue(content.payload, Map)
+
+    // currently supporting only interactive actions
+    if (payload.type != "interactive_message") {
+      throw new IllegalArgumentException("Unsupported Slack callback type: ${payload.type}")
+    }
+
+    if (!payload.callback_id || !payload.user?.name) {
+      throw new IllegalArgumentException("Slack callback_id and user not present. Cannot route the request to originating Spinnaker service.")
+    }
+
+    payload
+  }
+
+  @Override
+  InteractiveActionCallback parseInteractionCallback(Map content) {
+    log.debug("Received callback event from Slack of type ${content.type}")
+    // TODO: validate the X-Slack-Signature header (https://api.slack.com/docs/verifying-requests-from-slack)
+
+    Map payload = parseSlackPayload(content)
+
+    def (serviceId, callbackId) = payload.callback_id.split(":")
+    String user = payload.user.name
+    // TODO: get user e-mail from Slack API so it matches what we use in X-SPINNAKER-USER
+    new InteractiveActionCallback(
+      serviceId: serviceId,
+      messageId: callbackId,
+      user: user,
+      // TODO: buttons only for now -- check
+      actionPerformed: new Notification.ButtonAction(
+        name: payload.actions[0].name,
+        label: payload.actions[0].text,
+        value: payload.actions[0].value
+      )
+    )
+  }
+
+  @Override
+  void respondToCallback(Map content) {
+    Map payload = parseSlackPayload(content)
+    log.debug("Responding to Slack callback via ${payload.response_url}")
+
+    // Example: https://hooks.slack.com/actions/TSJ9N5Z3J/906732166263/3zDfYbq62ei4eNELNJcAaK0C
+    URI responseUrl = new URI(payload.response_url)
+
+    SlackHookService slackHookService = new RestAdapter.Builder()
```

Comment:
Instantiating a service client dynamically here is required because the URL is sent by Slack as part of the payload, i.e. we don't know it upfront (or even if the host/format won't change).

Reply:
Again, you should cache these services.

Reply:
How do I cache these if the URL is different every time? If we can assume the host and URI format are stable, then I can parameterize the call instead.